External/FEXCore/Source/Interface/Core/Dispatcher/X86Dispatcher.cpp

```diff
@@ -25,384 +26,338 @@
-
-    call(qword STATE_PTR(CpuStateFrame, Pointers.Common.ExitFunctionLink));
+  return L1Entry.HostCode;
+}
 
-    if (SignalSafeCompile) {
-      // Now restore the signal mask
-      // Living in the same location
-      // Backup rax
-      mov(r9, rax);
+/* ---------------------------------------------------------------------------------- */
+
+/*
+ * Call Core::ExitFunctionLink to obtain a continuation
+ * address (which can be either LoopTop or a JIT entry), 
+ * and continue with executing that code. 
+ * TODO: Core::ExitFunctionLink has a potential 
+ * short return path, for which we may want to omit the 
+ * two calls to sigprocmask:
+ */
+//static
+uint64_t X86ExitFunctionLinkerCode( FEXCore::Core::CpuStateFrame *Frame, uint64_t *record )
+{ 
+  uint64_t HostCode;
+
+  if (!SignalSafeCompile) {
+    // Just compile the code
+    HostCode = X86JITCore::ExitFunctionLink(Frame,record);
+  } else {
+    // When compiling code, mask all signals to reduce the chance of reentrant allocations
+    sigset_t ProcMask={(unsigned long)-1,(unsigned long)-1};
+    sigprocmask( SIG_SETMASK, &ProcMask, &ProcMask);
+    HostCode = X86JITCore::ExitFunctionLink(Frame,record);
+    sigprocmask( SIG_SETMASK, &ProcMask, NULL);
+  }
+  
+  return HostCode;
+}
 
-      mov(rdi, SIG_SETMASK);
-      mov(rsi, rsp);
-      mov(rdx, 0); // Don't care about result
-      mov(r10, 8);
-      mov(rax, SYS_rt_sigprocmask);
-      syscall();
+/* ---------------------------------------------------------------------------------- */
 
-      // Bring stack back
-      add(rsp, 16);
+/*
+ * Call the dispatcher animation framework defined above, 
+ * while providing a quick termination via a longjmp:
+ */
+static 
+void X86DispatchCode( FEXCore::Core::CpuStateFrame *Frame )
+{
+  if (!setjmp(Frame->EmuContext)) {
+      uint32_t aligner MIE_ALIGN(16);
+      Frame->ReturningStackLocation = reinterpret_cast<uintptr_t>(&aligner);
 
-      jmp(r9);
-    }
-    else {
-      jmp(rax);
-    }
+      X86CallCoreDispatchCodeAsm(Frame);
   }
+}
 
-  {
-    // Pause handler
-    ThreadPauseHandlerAddress = getCurr<uint64_t>();
-    L(ThreadPauseHandler);
+__attribute__((noreturn))
+//static 
+void X86ThreadStopHandlerCode( FEXCore::Core::CpuStateFrame *Frame )
+{
+  longjmp(Frame->EmuContext,1);
+}
 
-    mov(rdi, reinterpret_cast<uintptr_t>(CTX));
-    mov(rsi, STATE);
-    mov(rax, reinterpret_cast<uint64_t>(SleepThread));
+/* ---------------------------------------------------------------------------------- */
 
-    call(rax);
+    __attribute__((noreturn))
+    static void WaitUntilWeHitATestCase(){assert(0);}
 
-    // XXX: Unsupported atm
-    PauseReturnInstruction = getCurr<uint64_t>();
-    ud2();
-  }
 
-  {
-    CallbackPtr = getCurr<JITCallback>();
 
-    push(rbx);
-    push(rbp);
-    push(r12);
-    push(r13);
-    push(r14);
-    push(r15);
-    sub(rsp, 8);
+static 
+void X86ThreadPauseHandlerAddressCode( FEXCore::Core::CpuStateFrame *Frame )
```

Comment:
This is used for our GDB integration so it would be good to keep working.

Reply:
A GDB session will probably test this, right?